src/main/java/com/google/devtools/build/bfg/JavaSourceFileParser.java

```diff
@@ -124,11 +201,45 @@
         .forEach(
             (path, classes) -> {
               if (any(oneRulePerPackageRoots, p -> path.startsWith(p))) {
-                putOnCycle(classes, graph);
+                putOnCycle(classes, classToClass);
               }
             });
+  }
+
+  private String decideRuleKind(ReferencedClassesParser parser, Set<String> dependencies) {
+    CompilationUnit cu = parser.compilationUnit;
+    if (cu.types().isEmpty()) {
+      return "java_library";
+    }
+    AbstractTypeDeclaration topLevelClass = (AbstractTypeDeclaration) cu.types().get(0);
+    if ((topLevelClass.getModifiers() & Modifier.ABSTRACT) != 0) {
+      // Class is abstract, can't be a test.
+      return "java_library";
+    }
+
+    // JUnit 4 tests
+    if (parser.className.endsWith("Test") && dependencies.contains("org.junit.Test")) {
+      return "java_test";
+    }
+
+    for (MethodDeclaration m : filter(topLevelClass.bodyDeclarations(), MethodDeclaration.class)) {
+      if ((m.getModifiers() & Modifier.STATIC) != 0
```

Comment:
nit: for readability, can you make this block of boolean logic into a separate method? 

`private boolean isMainMethod(MethodDeclaration m) {...}`

Reply:
> Also made the method test if the return type is voi

:)